File: vn/src/main/java/org/sqlbuilder/vn/objects/Syntax.java
package org.sqlbuilder.vn.objects;

import org.sqlbuilder.common.*;
import org.sqlbuilder.vn.collector.VnSyntaxXmlProcessor;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.util.Comparator;
import java.util.Objects;

import javax.xml.parsers.ParserConfigurationException;

public class Syntax implements HasId, Insertable, Comparable<Syntax>
{
	public static final Comparator<Syntax> COMPARATOR = Comparator.comparing(Syntax::getSyntax);

	public static final SetCollector2<Syntax> COLLECTOR = new SetCollector2<>(COMPARATOR);

	private static final VnSyntaxXmlProcessor SYNTAX_XML_PROCESSOR = new VnSyntaxXmlProcessor();

	private final String syntax;

	// C O N S T R U C T O R

	public static Syntax make(final String syntax) throws IOException, SAXException, ParserConfigurationException
	{
		String syntax2 = syntax.replaceFirst("^<SYNTAX>", "").replaceFirst("</SYNTAX>$", "");
		try
		{
			syntax2 = SYNTAX_XML_PROCESSOR.process(syntax2);
		}
		catch (ParserConfigurationException | IOException | SAXException e)
		{
			System.err.println(syntax2);
			throw e;
		}
		var s = new Syntax(syntax2);
		COLLECTOR.add(s);
		return s;
	}

	private Syntax(final String syntax)
	{
		this.syntax = syntax;
	}

	// A C C E S S

	public String getSyntax()
	{
		return syntax;
	}

	@Override
	public Integer getIntId()
	{
		return COLLECTOR.apply(this);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Syntax that = (Syntax) o;
		return syntax.equals(that.syntax);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(syntax);
	}

	// O R D E R I N G

	@Override
	public int compareTo(final Syntax that)
	{
		return this.syntax.compareTo(that.syntax);
	}

	// I N S E R T

	@Override
	public String dataRow()
	{
		// id
		// syntax
		return String.format("'%s'", Utils.escape(syntax));
	}
}

File: vn/src/main/java/org/sqlbuilder/vn/objects/Syntax.kt
package org.sqlbuilder.vn.objects

import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.vn.collector.VnSyntaxXmlProcessor
import org.xml.sax.SAXException
import java.io.IOException
import java.util.*
import javax.xml.parsers.ParserConfigurationException

class Syntax private constructor(
    val syntax: String,
) : HasId, Insertable, Comparable<Syntax> {

    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Syntax
        return syntax == that.syntax
    }

    override fun hashCode(): Int {
        return Objects.hash(syntax)
    }

    // O R D E R I N G

    override fun compareTo(that: Syntax): Int {
        return syntax.compareTo(that.syntax)
    }

    // I N S E R T

    override fun dataRow(): String {
        return "'${escape(syntax)}'"
    }

    companion object {

        val COMPARATOR: Comparator<Syntax> = Comparator.comparing<Syntax, String> { it.syntax }

        @JvmField
        val COLLECTOR: SetCollector<Syntax> = SetCollector<Syntax>(COMPARATOR)

        private val SYNTAX_XML_PROCESSOR = VnSyntaxXmlProcessor()

        @JvmStatic
        @Throws(IOException::class, SAXException::class, ParserConfigurationException::class)
        fun make(syntax: String): Syntax {
            var syntax2 = syntax.replaceFirst("^<SYNTAX>".toRegex(), "").replaceFirst("</SYNTAX>$".toRegex(), "")
            try {
                syntax2 = SYNTAX_XML_PROCESSOR.process(syntax2)
                val s = Syntax(syntax2)
                COLLECTOR.add(s)
                return s
            } catch (e: ParserConfigurationException) {
                System.err.println(syntax2)
                throw e
            } catch (e: IOException) {
                System.err.println(syntax2)
                throw e
            } catch (e: SAXException) {
                System.err.println(syntax2)
                throw e
            }
        }
    }
}
